be/src/exprs/cast_expr_json.cpp

```diff
@@ -0,0 +1,199 @@
+        Slice slice = col.get_slice(_row);
+        _add_element(std::string_view(slice.data, slice.size));
+        return {};
+    }
+
+    Status do_visit(const StructColumn& col) {
+        if (_field_name.empty()) {
+            _builder->openObject();
+        } else {
+            _builder->add(_field_name, vpack::Value(vpack::ValueType::Object));
+        }
+        auto& names = col.field_names();
+        auto& columns = col.fields();
+        for (int i = 0; i < columns.size(); i++) {
+            auto name = names.size() > i ? names[i] : fmt::format("k{}", i);
+            auto& field_column = columns[i];
+            RETURN_IF_ERROR(cast_datum_to_json(field_column, _row, name, _builder));
+        }
+        if (!_builder->isClosed()) {
+            _builder->close();
+        }
+
+        return {};
+    }
+
+    Status do_visit(const MapColumn& col) {
+        if (_field_name.empty()) {
+            _builder->openObject();
+        } else {
+            _builder->add(_field_name, vpack::Value(vpack::ValueType::Object));
+        }
+        auto [map_start, map_size] = col.get_map_offset_size(_row);
+        auto key_col = col.keys_column();
+        auto val_col = col.values_column();
+
+        if (key_col->has_null()) {
+            return Status::NotSupported("key of Map should not be nullable");
+        }
+        if (key_col->is_nullable()) {
+            key_col = ColumnHelper::as_column<NullableColumn>(key_col)->data_column();
+        }
+
+        for (int i = map_start; i < map_start + map_size; i++) {
+            std::string name;
+            if (key_col->is_binary()) {
+                auto binary_col = ColumnHelper::as_column<BinaryColumn>(key_col);
+                name = binary_col->get_slice(i);
+            } else if (key_col->is_large_binary()) {
+                auto binary_col = ColumnHelper::as_column<LargeBinaryColumn>(key_col);
+                name = binary_col->get_slice(i);
+            } else {
+                // TODO(murphy) cast to string instead of debug
+                name = key_col->debug_item(i);
+            }
+            // VLOG(2) << "map key: " << key_col->debug_item(i) << " , name=" << name;
+            RETURN_IF_ERROR(cast_datum_to_json(val_col, i, name, _builder));
+        }
+
+        if (!_builder->isClosed()) {
+            _builder->close();
+        }
+        return {};
+    }
+
+    Status do_visit(const ArrayColumn& col) {
+        if (_field_name.empty()) {
+            _builder->openArray();
+        } else {
+            _builder->add(_field_name, vpack::Value(vpack::ValueType::Array));
+        }
+
+        auto [offset, size] = col.get_element_offset_size(_row);
+        auto elements = col.elements_column();
+        for (int i = offset; i < offset + size; i++) {
+            RETURN_IF_ERROR(cast_datum_to_json(elements, i, "", _builder));
+        }
+
+        if (!_builder->isClosed()) {
+            _builder->close();
+        }
+
+        return {};
+    }
+
```

Comment:
do_visit for DecimalV3Column is absent

use DecimalNonDecimalCast::decimal_to

Reply:
good suggestion